File: CPlusPlus/1004. 最大连续1的个数 III.cpp
#include<vector>

using namespace std;
class Solution {
public:
    int longestOnes(vector<int>& nums, int k) {
        int curr = 0;
        size_t left = 0;
        size_t right = 0;
        size_t res = 0;
        while(right <= nums.size()){
            if(curr <= k){
                if(right < nums.size() && nums[right] == 0) ++curr;
                ++right;
                if(curr <= k && right <= nums.size() && right - left > res){
                    res = right - left;
                }
            }
            else if(curr > k){
                if(nums[left] == 0) --curr;
                ++left;
            }
        }
        return res;
    }
};

File: Rust/1004. 最大连续1的个数 III.rs
use std::cmp;
impl Solution {
    pub fn longest_ones(nums: Vec<i32>, k: i32) -> i32 {
        let mut left = 0usize;
        let mut right = 0usize;
        let mut cnt = 0i32;
        let mut res = 0usize;
        while right < nums.len() {
            if cnt <= k {
                if nums[right] == 0 {
                    cnt = cnt + 1;
                }
                right = right + 1;
            } else {
                if left < right && nums[left] == 0 {
                    cnt = cnt - 1;
                }
                left = left + 1;
            }
            if cnt <= k {
                res = cmp::max(res, right - left);
            }
        }
        res as i32
    }
}
